### PR TITLE
Add E2E test for zero-config for Windows IIS .NET apps

### PR DESCRIPTION
**Pull request by pjanotti**
Creates an E2E test to validate that the deployment of the SFx.NET via the install script is working for .NET Framework apps hosted on Windows IIS.

### CODE REVIEW

**Review by atoulme — APPROVED** [review]

**Review by pmcollins — APPROVED** [review]
Nice!

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -487,6 +487,62 @@ jobs:
...
+    runs-on: windows-2022
+    needs: [ windows-msi ]
+    steps:
+      - name: Check out the codebase.
+        uses: actions/checkout@v3
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: ${{ env.GO_VERSION }}
```
Comment (jeffreyc-splunk) [inline .github/workflows/build-and-test.yml:500]:
```suggestion
        uses: actions/setup-go@v4
        with:
          go-version: ${{ env.GO_VERSION }}
          cache-dependency-path: '**/go.sum'
```

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -487,6 +487,62 @@ jobs:
             throw "Failed to uninstall fluentdwinsvc."
           }
 
+  windows-zeroconfig-e2e-test:
+    runs-on: windows-2022
```
Comment (jeffreyc-splunk) [inline .github/workflows/build-and-test.yml:491]: Is there any value in also testing on `windows-2019`?

Comment (pjanotti) [reply .github/workflows/build-and-test.yml:491]: In principle no: there shouldn't be any difference affecting the versions shipped with the .NET instrumentation, nor the app itself. There is always a possibility of some difference, but, it is extremely unlikely in this case.

File: tests/zeroconfig/windows/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Tests for Splunk Zero Configuration of .NET applications hosted on Windows IIS
+
+Tests in this directory validate that automatic instrumentation of
+.NET and .NET Framework applications hosted on Windows IIS are working under
+a deployment following the Zero Configuration procedures.
+
+## Requirements to run the tests locally
+
+The GitHub job `windows-zeroconfig-e2e-test` in the [`build-and-test`](../../../.github/workflows/build-and-test.yml)
```
Comment (rmfitzpatrick) [inline tests/zeroconfig/windows/README.md:9]: Having all the setup in ci logic can make maintainability and debugging difficult. Would it be reasonable to migrate to a ps1 script and invoke there?

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+resource_metrics:
+  - attributes:
+    scope_metrics:
+      - metrics:
+          - name: otelcol_exporter_sent_spans
```
Comment (rmfitzpatrick) [inline tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: Do you mind adding all the expected metrics including labels? `<ANY>` can be used as a value for dynamic ones where we mainly care about the attr presence/key.

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -487,6 +487,63 @@ jobs:
             throw "Failed to uninstall fluentdwinsvc."
           }
 
+  windows-zeroconfig-e2e-test:
```
Comment (rmfitzpatrick) [inline .github/workflows/build-and-test.yml:490]: I'm not sure this needs to run w/ `build-and-test` and think a new zero config workflow could be added that is limited to `internal/buildscripts/packaging/msi` and `instrumentation` content changes.

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -487,6 +487,63 @@ jobs:
...
+
+      - name: Build Docker compose
+        working-directory: tests/zeroconfig/windows/testdata/
+        run: |
+          docker compose build
+
+      - name: Run the test
+        working-directory: tests/zeroconfig/windows/
+        run: |
+          go test -timeout 5m -run ^TestWindowsIISInstrumentation$ -tags zeroconfig -v
```
Comment (rmfitzpatrick) [inline .github/workflows/build-and-test.yml:541]: `-tags zeroconfig` could be used if added to the test file so we don't need to filter w/ `-run`

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (jeffreyc-splunk) [reply .github/workflows/build-and-test.yml:490]: Only downside to moving to a new workflow is that it would require duplicating the build for all upstream deps (otelcol, agent bundle, and msi).

Alternatively, maybe we can try something like https://github.com/marketplace/actions/paths-changes-filter to only run the job when there are relevant changes?

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml (hunk 4 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: > Do you mind adding all the expected metrics including attributes*?

Most of the metrics are independent of the .NET instrumentation, or do you mean only the ones expected after the request to the instrumented application is done?

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (pjanotti) [reply .github/workflows/build-and-test.yml:490]: If adding the 3rd party action is acceptable it sounds good to me - this action seems reputable. Related Q.: is this enough to guarantee that it runs at least once for each release?

File: tests/zeroconfig/windows/README.md (hunk 3 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/README.md:9]: I personally prefer the separate action steps: typically you run a few of them and then repeat a specific one to debug/investigate. My guess is that the script will be run only in CI, for debug/investigation devs will copy and paste particular lines of the script. That said I will make the change so we can better judge it.

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (rmfitzpatrick) [reply .github/workflows/build-and-test.yml:490]: We could also run this for merges to main for general tracking coverage?

Comment (jeffreyc-splunk) [reply .github/workflows/build-and-test.yml:490]: If we want to use that action, I think we can add the action to the upstream `windows-msi` job and configure this job to always run for `main` or for relevant changes.  Maybe something like:
```
windows-msi:
  outputs:
    zeroconfig: ${{ steps.filter.outputs.zeroconfig }}
  steps:
    - name: Check out the codebase.
      uses: actions/checkout@v3
    - uses: dorny/paths-filter@v2
      id: filter
      with:
        filters: |
          zeroconfig:
            - path1
            - path2
            - ...
```
and then
```
windows-zeroconfig-e2e-test:
  needs: [ windows-msi ]
  if: ${{ github.ref_name == 'main' || needs.windows-msi.outputs.zeroconfig == 'true' }}
  steps: ...
```

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml (hunk 4 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: Yes, this is an internal collector metric. The SFx.NET metrics are disabled by default so I didn't want to rely on them.

File: tests/zeroconfig/windows/README.md (hunk 3 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/README.md:9]: After writing it, I liked it :) - let's see if it passes CI ...

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml (hunk 4 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: I have a story to add the resource spans to testutils, the plan is to switch this test to actually check for the spans.

File: tests/zeroconfig/windows/README.md (hunk 3 above)
Comment (rmfitzpatrick) [reply tests/zeroconfig/windows/README.md:9]: Thanks!

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml (hunk 4 above)
Comment (rmfitzpatrick) [reply tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: Adding resource spans checks sounds great though it took me a bit to understand the evaluated flow here. You're installing two collectors, one for zero config w/ ingest url configured to be the signalfx receiver of the second collector, which is configured to report to the test's otlp sink?

File: tests/zeroconfig/windows/testdata/Dockerfile.pipeline.collector [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+# WARNING: Passing access token as a build argument is to facilitate manual validation of this Dockerfile. It shouldn't be used in production.
+ARG verify_access_token=false
+ARG access_token=some_token
+ENV VERIFY_ACCESS_TOKEN=$verify_access_token
+ENV ACCESS_TOKEN_TMP=$access_token
+RUN `
+    $token = $Env:ACCESS_TOKEN_TMP; `
+    $collector_msi_path = (dir "c:\setup\splunk-otel-collector-*.msi").FullName; `
+    c:\setup\install.ps1 -access_token $token -msi_path $collector_msi_path -with_fluentd $false -ingest_url "http://10.1.1.1:4318"
```
Comment (rmfitzpatrick) [inline tests/zeroconfig/windows/testdata/Dockerfile.pipeline.collector:21]: mind configuring it to use a reserved/unused address in the off chance something leads it to report to the otlp sink directly?

```suggestion
    c:\setup\install.ps1 -access_token $token -msi_path $collector_msi_path -with_fluentd $false -ingest_url "http://192.0.2.1:12345"
```

File: tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml (hunk 4 above)
Comment (pjanotti) [reply tests/zeroconfig/windows/testdata/expected_resource_metrics.yaml:5]: yes, I tried to document that on the test code, may require some re-wording to make that clear... see https://github.com/signalfx/splunk-otel-collector/pull/2901/files#diff-164dad5cd795a3f716eea69cefc15b2062e4d9d766cfa9ee103ae943716a40f1R41

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (pjanotti) [reply .github/workflows/build-and-test.yml:490]: I guess this is the last pending thing to merge this PR. I will take a look in the afternoon PT.

Comment (pjanotti) [reply .github/workflows/build-and-test.yml:508]: I opted to have a dedicated job instead of piggybacking on `windows-msi` since the job is very light weight, but, no strong opinion about that, see https://github.com/signalfx/splunk-otel-collector/pull/2901/commits/2ce7fb89dca76f1569d7968a2857d99b9877eeda

File: tests/zeroconfig/windows/windows_iis_test.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	// New-NetFirewallRule -DisplayName 'zc-iis-test' -Direction Inbound -LocalAddress 10.1.1.1 -LocalPort 4318 -Protocol TCP -Action Allow -Profile Any
+	//
+	// The command to remove the rule is:
+	//
+	// Remove-NetFirewallRule -DisplayName 'zc-iis-test'
+	//
+	// The OTLP also requires the Docker compose network to be created before it is initialized, since
+	// the address is part of that network.
+	//
+	otlp, err := testutils.NewOTLPReceiverSink().WithEndpoint("10.1.1.1:4318").Build()
```
Comment (rmfitzpatrick) [inline tests/zeroconfig/windows/windows_iis_test.go:68]: Thanks for revealing a pain point. Ideally this would be possible w/ the testcase helpers but hardcoding the endpoint complicates it. Will be sure to add an option for this and update in the future.

File: .github/workflows/build-and-test.yml [hunk 9]
```diff
@@ -487,6 +487,55 @@ jobs:
...
+      changed: ${{ steps.filter.outputs.changed }}
+    steps:
+    - uses: dorny/paths-filter@v2
+      id: filter
+      with:
+        filters: |
+          changed:
+            - 'internal/buildscripts/packaging/installer/install.ps1'
+            - 'internal/buildscripts/packaging/msi/**'
+            - 'pkg/signalfx-agent/bundle/**'
```
Comment (jeffreyc-splunk) [inline .github/workflows/build-and-test.yml:506]: Should the tests also be included?
```suggestion
            - 'pkg/signalfx-agent/bundle/**'
            - 'tests/zeroconfig/**'
```

Comment (pjanotti) [reply .github/workflows/build-and-test.yml:506]: Good point @jeffreyc-splunk pending...

**Review by rmfitzpatrick — APPROVED** [review]
Thanks!

**Review by jeffreyc-splunk — APPROVED** [review]
